Reduce FuncDeclaration size

RazvanN7 opened the pull request:
by grouping contract-related member fields.

RazvanN7 committed:
Reduce FuncDeclaration size

maxhaton commented on compiler/src/dmd/func.d:
```diff
@@ -258,22 +258,29 @@ extern (C++) struct Ensure
-/***********************************************************
- */
-extern (C++) class FuncDeclaration : Declaration
+// Most function don't have contracts, so save memory by grouping
```
```suggestion
// Most functions don't have contracts, so save memory by grouping
```

Also use a multiline doc comment

maxhaton commented on compiler/src/dmd/func.d:
```diff
@@ -258,22 +258,29 @@ extern (C++) struct Ensure
-
     Expressions* fdrequireParams;       /// argument list for __require
     Expressions* fdensureParams;        /// argument list for __ensure
+}
+
+/***********************************************************
+ */
+extern (C++) class FuncDeclaration : Declaration
```
This diff looks unnecessary. The class declaration should be unaffected

RazvanN7 replied:
Github is just being weird here, I didn't modify the class definition.

maxhaton commented on compiler/src/tests/cxxfrontend.cc:
```diff
@@ -964,8 +964,8 @@ class MiniGlueVisitor : public Visitor
         sym->statement->accept(this);
-        if (sym == func->returnLabel && func->fensure != NULL)
-            func->fensure->accept(this);
+        if (sym == func->returnLabel && func->fensure() != NULL)
```
This is probably a breaking change for ldc and gdc glue code?

@ibuclaw @kinke ?

maxhaton reviewed (changes requested):
Previous approve was by mistake, sorry.

maxhaton commented:
frontend.h!

maxhaton commented:
frontend.h!

WalterBright commented on compiler/src/dmd/func.d:
```diff
@@ -403,6 +410,75 @@ extern (C++) class FuncDeclaration : Declaration
+            if (!frs && !contracts)
+                return null;
+            return getContracts().frequires = frs;
+        }
```
With a lot of duplicated code, a mixin template might be more suitable. Or a string mixin.

RazvanN7 commented:
@maxhaton I addressed your concerns.

@WalterBright Yes, that is what I thought initially, but I wanted to get it right first (as string mixins are a bit cumbersome to debug) and see if it is something that will get accepted. I am using a string mixin now.

RazvanN7 commented:
@maxhaton is this good to go?

maxhaton commented:
I'm happy with it, I assume Iain and Martin can work around it but wait for confirmation from them (ldc is currently lagging behind dmd, this isn't dramatic breakage if it is indeed breakage but just keep it in mind)

kinke commented:
Yeah this is just regular breakage which the C++ compilers handle quite well usually; additionally, I doubt these members are used in the glue layer.

RazvanN7 commented:
@kinke Is there anything I can do to make things easier for you?
